src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -213,18 +213,47 @@ internal JsonTypeInfo(Type type, JsonConverter converter, JsonSerializerOptions
         /// It is required that added <see cref="JsonPropertyInfo"/> entries are unique up to <see cref="JsonPropertyInfo.Name"/>,
         /// however this will only be validated on serialization, once the metadata instance gets locked for further modification.
         /// </remarks>
-        public IList<JsonPropertyInfo> Properties => _properties ??= new(this);
-        private JsonPropertyInfoList? _properties;
+        public IList<JsonPropertyInfo> Properties => PropertyList;
 
-        internal void SortProperties()
+        internal JsonPropertyInfoList PropertyList
         {
-            Debug.Assert(!IsConfigured);
-            Debug.Assert(_properties != null && _properties.Count > 0);
+            get
+            {
+                return _properties ?? CreatePropertyList();
+                JsonPropertyInfoList CreatePropertyList()
+                {
+                    var list = new JsonPropertyInfoList(this);
+                    if (_sourceGenDelayedPropertyInitializer is { } propInit)
+                    {
+                        // .NET 6 source gen backward compatibility -- ensure that the
+                        // property initializer delegate is invoked lazily.
+                        JsonMetadataServices.PopulateProperties(this, list, propInit);
+                    }
 
-            _properties.SortProperties();
-            PropertyCache?.List.StableSortByKey(static propInfo => propInfo.Value.Order);
+                    JsonPropertyInfoList? result = Interlocked.CompareExchange(ref _properties, list, null);
+                    _sourceGenDelayedPropertyInitializer = null;
```

Comment:
What is the impact of nulling this out?  Is that just for hygiene / not holding on to things unnecessarily, or is there a functional impact?  Just wondering from a synchronization perspective.